cpp/src/arrow/compute/CMakeLists.txt

```diff
@@ -114,6 +114,47 @@ function(ADD_ARROW_COMPUTE_TEST REL_TEST_NAME)
                  ${ARG_UNPARSED_ARGUMENTS})
 endfunction()
 
+# This function is used to add a custom main to the benchmarks in order
+# to initialize the compute kernels registry before running them.
+# This is necessary for benchmarks that use compute kernels that are not
+# part of libarrow.
+# It will also link the compute libraries to the benchmark target.
+function(ADD_ARROW_COMPUTE_BENCHMARK REL_TEST_NAME)
+  set(options)
+  set(one_value_args PREFIX)
+  set(multi_value_args EXTRA_SOURCES)
+  cmake_parse_arguments(ARG
+                        "${options}"
+                        "${one_value_args}"
+                        "${multi_value_args}"
+                        ${ARGN})
+  if(ARG_PREFIX)
+    set(PREFIX ${ARG_PREFIX})
+  else()
+    set(PREFIX "arrow-compute")
+  endif()
+  if(ARG_EXTRA_SOURCES)
+    set(EXTRA_SOURCES ${ARG_EXTRA_SOURCES})
+  else()
+    set(EXTRA_SOURCES "${CMAKE_CURRENT_FUNCTION_LIST_DIR}/main_compute_benchmark.cc")
```

Comment:
Shouldn't this always be added to `EXTRA_SOURCES`? Right now, isn't it only added when `ARG_EXTRA_SOURCES` is blank?

Reply:
that's a good point, my initial thought was that someone calling the function might want to override this with their own `EXTRA_SOURCES` even providing their own main function but I am happy to be less flexible and force the addition of this source for any `add_arrow_compute_benchmark` call.

Reply:
to be honest it makes more sense with the description of the function so I pushed https://github.com/apache/arrow/pull/46922/commits/4d3856ffd07e2ecb6328d4ef7d348fbae31c0801 which adds the file unconditionally